extmod/modframebuf.c

```diff
@@ -174,11 +174,107 @@ STATIC mp_obj_t framebuf1_text(size_t n_args, const mp_obj_t *args) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(framebuf1_text_obj, 4, 5, framebuf1_text);
 
+// Fill a rectangle.
+STATIC void framebuf1_fillbox(const mp_obj_framebuf1_t *fb, int x, int y, int width, int height, uint8_t color) {
+    if (x < 0 || y < 0 || y >= fb->height || x >= fb->width || height <= 0 || width <= 0) {
+        // No operation needed.
+        return;
+    }
+
+    // clip to the display
+    x = MAX(MIN(x, fb->width), 0);
+    y = MAX(MIN(y, fb->height), 0);
+    int xend = MIN(fb->width, x + width);
+    int yend = MIN(fb->height, y + height);
+    int start_row = y >> 3;
+    int end_row = yend >> 3;
+    uint8_t start_mask, end_mask;
+
+    if (color) {
+        if (end_row == start_row) {
+            start_mask = ((1 << (yend - y)) - 1) << (y & 0x07);
+            for(int j = x; j < xend; ++j) {
+                fb->buf[start_row * fb->stride + j] |= start_mask;
+            }
+        } else {
+            start_mask = 0xff << (y & 0x07);
+            end_mask = ~(0xff << (yend & 0x07));
+            for(int j = x; j < xend; ++j) {
+                fb->buf[start_row * fb->stride + j] |= start_mask;
+                for (int i = start_row + 1; i < end_row; ++i) {
+                    fb->buf[i * fb->stride + j] = 0xff;
+                }
+                fb->buf[end_row * fb->stride + j] |= end_mask;
+            }
+        }
+    } else {
+        if (end_row == start_row) {
+            start_mask = ((1 << (yend - y)) - 1) << (y & 0x07);
+            for(int j = x; j < xend; ++j) {
+                fb->buf[start_row * fb->stride + j] &= ~start_mask;
+            }
+        } else {
+            start_mask = 0xff << (y & 0x07);
+            end_mask = ~(0xff << (yend & 0x07));
+            for(int j = x; j < xend; ++j) {
+                fb->buf[start_row * fb->stride + j] &= ~start_mask;
+                for (int i = start_row + 1; i < end_row; ++i) {
+                    fb->buf[i * fb->stride + j] = 0xff;
+                }
+                fb->buf[end_row * fb->stride + j] &= ~end_mask;
+            }
+        }
+    }
+}
+
+// Draw a horizontal line.
+STATIC mp_obj_t framebuf1_hline(size_t n_args, const mp_obj_t *args) {
+    mp_obj_framebuf1_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_int_t x = mp_obj_get_int(args[1]);
+    mp_int_t y = mp_obj_get_int(args[2]);
+    mp_int_t width = mp_obj_get_int(args[3]);
+    mp_int_t color = mp_obj_get_int(args[4]);
+
+    framebuf1_fillbox(self, x, y, width, 1, color);
```

Comment:
I think you can reduce code size by getting framebuf1_fillbox to unpack the Python objects (and instead of passing 1 here pass MP_OBJ_NEW_SMALL_INT(1)).